yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -451,8 +452,20 @@ private[spark] class ApplicationMaster(
   private def startUserClass(): Thread = {
     logInfo("Starting the user JAR in a separate Thread")
     System.setProperty("spark.executor.instances", args.numExecutors.toString)
-    val mainMethod = Class.forName(args.userClass, false,
-      Thread.currentThread.getContextClassLoader).getMethod("main", classOf[Array[String]])
+
+    val classpath = Client.getUserClasspath(null, sparkConf)
+    val urls = classpath.map { entry =>
+      new URL("file:" + new File(entry.getPath()).getAbsolutePath())
+    }
+    val userClassLoader =
+      if (Client.isUserClassPathFirst(sparkConf, true)) {
+        new ChildExecutorURLClassLoader(urls, Utils.getContextOrSparkClassLoader)
+      } else {
+        new ExecutorURLClassLoader(urls, Utils.getContextOrSparkClassLoader)
+      }
+
+    val mainMethod = userClassLoader.loadClass(args.userClass)
```

Comment:
What's the difference between using loader.loaderClass as we do here vs. Class.forName elsewhere?  Can we keep them consistent?